extmod/btstack/modbluetooth_btstack.c

```diff
@@ -85,6 +83,148 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(uint16_t uuid16, const uint8_t *uu
     return result;
 }
 
+// Notes on supporting background ops (i.e. an attempt to gatts_notify while and existing
+// notification is in progress):
+
+// GATTS Notify/Indicate (att_server_notify/indicate)
+// * When available, copies buffer immediately.
+// * Otherwise fails with BTSTACK_ACL_BUFFERS_FULL
+// * Use att_server_request_to_send_notification/indication to get callback
+//   * Takes btstack_context_callback_registration_t (and takes ownership) and conn_handle.
+//   * Callback is invoked with just the context member of the btstack_context_callback_registration_t
+
+// GATTC Write without response (gatt_client_write_value_of_characteristic_without_response)
+// * When available, copies buffer immediately.
+// * Otherwise, fails with GATT_CLIENT_BUSY.
+// * Use gatt_client_request_can_write_without_response_event to get callback
+//   * Takes btstack_packet_handler_t (function pointer) and conn_handle
+//   * Callback is involved, use gatt_event_can_write_without_response_get_handle to get the conn_handle (no other context)
+//   * There can only be one pending gatt_client_request_can_write_without_response_event (otherwise we fail with EALREADY).
+
+// GATTC Write with response (gatt_client_write_value_of_characteristic)
+// * When peripheral is available, takes ownership of buffer.
+// * Otherwise, fails with GATT_CLIENT_IN_WRONG_STATE (we fail the operation).
+// * Raises GATT_EVENT_QUERY_COMPLETE to the supplied packet handler.
+
+// For notify/indicate/write-without-response that proceed immediately, nothing extra required.
+// For all other cases, buffer needs to be copied and protected from GC.
+// For notify/indicate:
+//  * btstack_context_callback_registration_t:
+//     * needs to be malloc'ed
+//     * needs to be protected from GC
+//     * context arg needs to point back to the callback registration so it can be freed and un-protected
+// For write-without-response
+//  * only the conn_handle is available in the callback
+//  * so we need a queue of conn_handle->(value_handle, copied buffer)
+
+// Pending operation types.
+enum {
+    // Queued for sending when possible.
+    MP_BLUETOOTH_BTSTACK_PENDING_NOTIFY, // Waiting for context callback
+    MP_BLUETOOTH_BTSTACK_PENDING_INDICATE, // Waiting for context callback
+    MP_BLUETOOTH_BTSTACK_PENDING_WRITE_NO_RESPONSE, // Waiting for conn handle
+    // Hold buffer pointer until complete.
+    MP_BLUETOOTH_BTSTACK_PENDING_WRITE, // Waiting for write done event
+};
+
+// Pending operation:
+//  - Holds a GC reference to the copied outgoing buffer.
+//  - Provides enough information for the callback handler to execute the desired operation.
+struct _mp_btstack_pending_op_t {
+    mp_btstack_pending_op_t *next; // Must be first field to match btstack_linked_item.
```

Comment:
Following BTstack, and to make it clear this struct is a valid linked-list, this should probably be `btstack_linked_item_t *next;`